File: scripts/filter_cbd_languages.R
#!/usr/bin/env Rscript
# filter_cbd_languages.R
# Filter glottolog_data.csv to keep only languages spoken in the 196 CBD Party countries.
#
# The cbd_parties.csv file contains 224 rows: 196 sovereign Parties (including the EU
# as one Party covering 27 member states), plus non-Parties (Holy See, USA). The EU
# member states appear twice — once under their own entry and once under the EU — so
# we deduplicate by ISO_A2 code. Two rows (Côte d'Ivoire and Türkiye) have blank
# ISO_A2 codes; we manually assign them (CI, TR).
#
# This script produces the same cbd_party_languages.csv as the Python version but
# correctly resolves to the full 196 Parties.

library(tidyverse)
library(janitor)

# ── Paths ──────────────────────────────────────────────────────────────────────

project_root <- here::here()

cbd_file      <- file.path(project_root, "data", "cbd_parties.csv")
glottolog_file <- file.path(project_root, "data-raw", "glottolog_data.csv")
output_file   <- file.path(project_root, "data", "cbd_party_languages.csv")

# ── Manual fixes for rows with missing ISO codes ───────────────────────────────

iso2_fixes <- c(
  "Côte d'Ivoire" = "CI",
  "Türkiye"       = "TR"
)

# ── Load and clean CBD parties ─────────────────────────────────────────────────

cbd_raw <- read_csv(cbd_file, show_col_types = FALSE) %>%
  janitor::clean_names()

cbd_parties <- cbd_raw %>%
  # Only keep rows where Party date is present (i.e. ratified/acceded)
  filter(!is.na(party) & str_trim(party) != "") %>%
  # Fix missing ISO_A2 codes
  mutate(
    iso_a2 = str_trim(iso_a2),
    iso_a2 = if_else(
      iso_a2 == "" | is.na(iso_a2),
      recode(country, !!!iso2_fixes),
      iso_a2
    )
  ) %>%
  # De-duplicate: EU member states appear both individually and under the EU entry
  distinct(iso_a2, .keep_all = TRUE) %>%
  filter(!is.na(iso_a2) & iso_a2 != "") %>%
  select(country, iso_a2, iso_a3, party)

cat(sprintf("Loaded %d CBD Party ISO_A2 codes\n", nrow(cbd_parties)))

# ── Load and filter Glottolog data ─────────────────────────────────────────────

cbd_codes <- unique(cbd_parties$iso_a2)

glottolog <- read_csv(glottolog_file, show_col_types = FALSE) %>%
  # Only language-level entries
  filter(core.level == "language")

cat(sprintf("Found %d language-level entries in Glottolog\n", nrow(glottolog)))

# Explode country codes and keep rows where at least one country is a CBD Party
cbd_languages <- glottolog %>%
  # Split pipe-separated country codes into individual rows
  mutate(
    core_countries_raw = core_countries,
    core_countries_list = str_split(core_countries, "\\s*\\|\\s*")
  ) %>%
  unnest(core_countries_list) %>%
  mutate(iso_a2 = str_trim(core_countries_list)) %>%
  filter(iso_a2 != "") %>%
  # Keep only rows where the country is a CBD Party
  filter(iso_a2 %in% cbd_codes) %>%
  # Rebuild the pipe-separated country string with only CBD Party codes
  group_by(across(-core_countries_list)) %>%
  summarise(
    core_countries = str_c(sort(unique(iso_a2)), collapse = " | "),
    .groups = "drop"
  ) %>%
  ungroup() %>%
  # Remove the temporary columns we added
  select(-core_countries_raw, -iso_a2)

cat(sprintf(
  "Kept %d languages, skipped %d (not in any CBD Party country)\n",
  nrow(cbd_languages),
  nrow(glottolog) - nrow(cbd_languages)
))

# ── Write output ───────────────────────────────────────────────────────────────

write_csv(cbd_languages, output_file)
cat(sprintf("Output written to %s\n", output_file))

File: scripts/filter_cbd_languages.py
#!/usr/bin/env python3
"""
Filter glottolog_data.csv to keep only languages spoken in CBD Party countries.
- Reads CBD party ISO_A2 codes from cbd_parties2.csv
- For each language in glottolog_data.csv, checks if any country code in core.countries
  matches a CBD Party ISO_A2 code
- Outputs matching rows to cbd_party_languages.csv
"""

import csv
import io
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent
CBD_FILE = PROJECT_ROOT / "data" / "cbd_parties.csv"
GLOTTOLOG_FILE = PROJECT_ROOT / "data-raw" / "glottolog_data.csv"
OUTPUT_FILE = PROJECT_ROOT / "data" / "cbd_party_languages.csv"


def load_cbd_party_codes(filepath):
    """Load the set of ISO_A2 codes for CBD Party countries.

    Only includes countries that have actually ratified/acceded to the CBD
    (i.e. have a Party date). Non-Parties like Holy See and USA are excluded.
    """
    codes = set()
    with open(filepath, "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            # The last column is the Party date; if empty, the country is not a Party
            party_date = row.get("Party", "").strip()  # last column = Party date
            code = row["ISO_A2"].strip()
            if code and party_date:
                codes.add(code)
    return codes


def filter_glottolog_by_cbd(cbd_codes, glottolog_path, output_path):
    """Filter glottolog rows where any core.countries code is in the CBD set."""
    with open(glottolog_path, "r", encoding="utf-8") as fin, \
         open(output_path, "w", encoding="utf-8", newline="") as fout:

        # Read header to find column indices
        header_line = fin.readline()
        # Use csv reader for proper parsing since some fields contain commas in quotes
        fin.seek(0)

        reader = csv.DictReader(fin)
        header = reader.fieldnames
        countries_col = "core.countries"

        writer = csv.DictWriter(fout, fieldnames=header)
        writer.writeheader()

        kept = 0
        skipped = 0
        for row in reader:
            raw = row.get(countries_col, "") or ""
            # Parse pipe-separated country codes, strip whitespace
            if raw.strip() == "":
                skipped += 1
                continue
            lang_codes = {c.strip() for c in raw.split("|")}
            # Keep if any language country code is a CBD party
            if lang_codes & cbd_codes:
                # Strip non-CBD country codes from core.countries
                filtered_codes = sorted(c.strip() for c in raw.split("|") if c.strip() in cbd_codes)
                row[countries_col] = " | ".join(filtered_codes)
                writer.writerow(row)
                kept += 1
            else:
                skipped += 1

    return kept, skipped


if __name__ == "__main__":
    cbd_codes = load_cbd_party_codes(CBD_FILE)
    print(f"Loaded {len(cbd_codes)} CBD Party ISO_A2 codes")
    print(f"Codes: {sorted(cbd_codes)}")

    kept, skipped = filter_glottolog_by_cbd(cbd_codes, GLOTTOLOG_FILE, OUTPUT_FILE)
    print(f"Kept {kept} languages, skipped {skipped} (not in any CBD Party country)")
    print(f"Output written to {OUTPUT_FILE}")
